framework/src/messaging/messaging_sndinternal.c

```diff
@@ -0,0 +1,221 @@
+			return ERROR;
+		}
+	}
+
+	send_packet = (char *)MSG_ALLOC(send_size);
+	if (send_packet == NULL) {
+		msgdbg("[Messaging] send fail : out of memory for including header.\n");
+		return ERROR;
+	}
+
+	/* Send packet is like below.
+	 * +-----------------------------------------+
+	 * | sender_pid | msg type | message        |
+	 * +-----------------------------------------+
+	 */
+
+	/* Add data header for sender pid. */
+	sender_pid = getpid();
+	memcpy(send_packet, &sender_pid, sizeof(pid_t));
+	/* Add data header for send type. */
+	if (msg_type == MSG_SEND_NOREPLY) {
+		send_type = MSG_REPLY_NO_REQUIRED;
+	} else {
+		send_type = MSG_REPLY_REQUIRED;
+	}
+	memcpy(send_packet + sizeof(pid_t), &send_type, sizeof(int));
+	/* Copy the real send message. */
+	memcpy(send_packet + MSG_HEADER_SIZE, send_data->msg, send_data->msglen);
+
+	ret = mq_send(mqdes, (char *)send_packet, send_size, send_data->priority);
+	if (ret != OK) {
+		msgdbg("message send fail : send fail %d.\n", errno);
+		MSG_FREE(send_packet);
+		mq_close(mqdes);
+		mq_unlink(port_name);
+		return ERROR;
+	}
+
+	MSG_FREE(send_packet);
+	mq_close(mqdes);
+	return ret;
+}
+
+static void messaging_init_recv_arr(int *arr)
+{
+	int arr_idx;
+	for (arr_idx = 0; arr_idx < CONFIG_MESSAGING_RECV_LIST_SIZE; arr_idx++) {
+		arr[arr_idx] = MSG_RECV_NOT_INIT;
+	}
+}
+/****************************************************************************
+ * Name : messaging_send_internal
+ * 
+ * Description:
+ *  This function checks how many receivers are waiting and send message to them.
+ *
+ * Input Parameters:
+ *  port_name : The message port name to send
+ *  msg       : The message to be sent
+ *  msglen    : The length of message to be sent
+ *  priority  : A non-negative integer that specifies the priority of this message
+ * 
+ * Return Value:
+ *  On success, the number of waiting receivers is returned.
+ *  On failure, -1 (ERROR) is returned.
+ ****************************************************************************/
+int messaging_send_internal(const char *port_name, int msg_type, msg_send_data_t *send_data, msg_callback_info_t *cb_info)
+{
+	int recv_idx;
+	int ret = ERROR;
+	int read_status = MSG_READ_YET;
+	int recv_arr[CONFIG_MESSAGING_RECV_LIST_SIZE];
+	char *private_portname;
+	int recv_cnt;
+
+	/* Check that how many receivers are waiting. */
+	while (read_status != MSG_READ_ALL) {
+		(void)messaging_init_recv_arr(recv_arr);
+		read_status = messaging_handle_data(MSG_HANDLE_READ, port_name, recv_arr, &recv_cnt);
+		if (read_status == ERROR) {
+			msgdbg("[Messaging] send fail : getting receivers info error.\n");
+			return ERROR;
+		}
+
+		if (msg_type != MSG_SEND_MULTI && recv_cnt > 1) {
+			msgdbg("[Messaging] send fail : too many receivers are waiting unicast.\n");
+			return ERROR;
+		}
+
+		/* Send message to each receivers. */
+		for (recv_idx = 0; recv_idx < recv_cnt; recv_idx++) {
+			if (recv_arr[recv_idx] != MSG_RECV_NOT_INIT) {
```

Comment:
It is acceptable but if depth is too deep, you have to chec condition.
Here there is only if case, then below is much better to save your eye.
```
if (A == B) {
    continue;
}
MSG_AS....